src/crimson/net/SocketConnection.cc

```diff
@@ -314,8 +314,13 @@ seastar::future<> SocketConnection::send(MessageRef msg)
 
 seastar::future<> SocketConnection::close()
 {
-  get_messenger()->unregister_conn(this);
-  return seastar::when_all(in.close(), out.close()).discard_result();
+  return seastar::when_all(
```

Comment:
i think we should close the `input_stream` and `out_stream` first, otherwise, we could have following sequence:

1. in.close(); out.close()
1. a user of this messenger wants to send a message to an entity, and `loopup_conn()` returns a closed connection
1. `get_messenger()->unregister_conn(conn)`
1. the user uses this invalid connection for communication with the entity

Reply:
we should `unregister_conn()` before closing the streams for the reason @tchaikov mentioned. i think we should also change the connection state to `state_t::closed` before closing streams

Reply:
> No one is holding the connection (because `2.2` & `3.3`), so it starts to destruct at `SocketConnection.cc#L43`

i don't get this. as 2.3 puts

> task being scheduled, leaving `out_stream.close()` not being executed ...

which implies that the continuation of `connections.clear()` scheduled by `SocketMessenger::shutdown()` is not returned yet. how come the dtor of connection is called? it won't be called until all connections managed by that messenger are `close()`'ed . and `SocketConnection::close()` is considered completed only if `seastar::when_all(in.close(), out.close()).discard_result()` returns.

Reply:
To be clear, this bug happens because `~SocketConnection()` executes before `out.close()`. Which means at least 3 places should not hold `ConnectionRef` before `out.close()`:
a. The anonymous function to close the `ConnectionRef` is returned: https://github.com/ceph/ceph/blob/master/src/crimson/net/SocketMessenger.cc#L154
b. The `ConnectionRef` is erased from `SocketMessenger::connections`: https://github.com/ceph/ceph/blob/master/src/crimson/net/SocketMessenger.cc#L192
c. `SocketMessenger::dispatch(conn)` is returned for the `ConnectionRef`: https://github.com/ceph/ceph/blob/master/src/crimson/net/SocketMessenger.cc#L48

Then your question:
Yes, `connection.clear()` is not executed yet. And yes, `SocketMessenger::shutdown()` is not returned yet. But these *don't* imply that the anonymous function `a.` isn't returned yet. Actually, it is returned and the `ConnectionRef` is released before `connection.clear()` and is even before `out.close()`. They are asynchronous continuations attached to the return operation, only their futures are constructed right after returning.

I think https://github.com/scylladb/seastar/blob/master/doc/tutorial.md#sharing-ownership-reference-counting is telling a similar story: "So we need to ensure that something does hold on to another copy of f until the asynchronous method call completes." And its solution also works here:

REPLACE LINE
https://github.com/ceph/ceph/blob/master/src/crimson/net/SocketMessenger.cc#L154
> return conn.second->close();

BY
> return conn.second->close().finally([conn=std::move(conn.second)] {});

And it works without the need to move `get_messenger()->unregister_conn(this);` .

Reply:
please note `SocketMessenger::connections` holds strong references of `Connection`. so the the dtor of `connections` should not be called before `connections.clear()`.

Reply:
makes sense, thanks! i think we should fix this inside of `SocketConnection::close()` itself though, so that it applies to other callers too. how about this?
```c++
seastar::future<> SocketConnection::close()
{
  // unregister_conn() drops a reference, so hold another until completion
  auto cleanup = [conn = ConnectionRef(this)] {};
 
  get_messenger()->unregister_conn(this);
 
  return seastar::when_all(in.close(), out.close())
      .discard_result()
      .finally(std::move(cleanup));
}
```

Reply:
After `conn.second->close()` returned and before `connections.clear()`, `get_messenger()->unregister_conn(this)` is executed inside `conn.second->close()` already. This function erases the `connection` from `SocketMessenger::connections`.

If my understanding is correct, then I don't understand why the dtor of the `connection` can not be called before `connections.clear()`, since `SocketMessenger::connections` doesn't hold the reference of the closed and erased `connection`.

And after the fix mentioned, `Assertion '!_in_batch' failed` failure won't happen, seems it does make differences.